### ISSUE TITLE
Unable to deserialise headers and queryStringParameters from incoming (local) API Gateway request

### ISSUE THREAD

**Issue by patrickherrera** [awslabs/aws-serverless-java-container#243]
* Framework version: aws-serverless-java-container-core:1.3
* Implementations: all

## Scenario
I am using the 'sam local' (SAM CLI, version 0.11.0) tool to develop and test a lambda.
I am effectively following this guide: https://github.com/micronaut-projects/micronaut-aws/tree/master/examples/api-proxy-example to perform the integration

I am running `sam local start-api` to startup an API Gateway and Lambda container.
## Expected behaviour
I am making a request and passing query params. For example
`curl -v http://127.0.0.1:3000/landing?someQueryKey=someQueryValue`

I am expecting that my request will be transformed by the API gateway into a Proxy Request payload, passed to the Lambda and then deserialised into an `com.amazonaws.serverless.proxy.model.AwsProxyRequest` object.  
This is mostly working, but headers and queryStringParameters are being lost.
## Actual behaviour
I have added logging with the handler method to log the raw InputStream and also the deserialised AwsProxyRequest:

Raw Input Stream:
`{"httpMethod": "GET", "body": null, "resource": "/{proxy+}", "requestContext": {"resourceId": "123456", "apiId": "1234567890", "resourcePath": "/{proxy+}", "httpMethod": "GET", "requestId": "c6af9ac6-7b61-11e6-9a41-93e8deadbeef", "accountId": "123456789012", "stage": "prod", "identity": {"apiKey": null, "userArn": null, "cognitoAuthenticationType": null, "caller": null, "userAgent": "Custom User Agent String", "user": null, "cognitoIdentityPoolId": null, "cognitoAuthenticationProvider": null, "sourceIp": "127.0.0.1", "accountId": null}, "extendedRequestId": null, "path": "/{proxy+}"}, "queryStringParameters": {"someQueryKey": "someQueryValue"}, "headers": {"Host": "127.0.0.1:3000", "User-Agent": "curl/7.61.1", "Accept": "*/*", "X-Forwarded-Proto": "http", "X-Forwarded-Port": "3000"}, "pathParameters": {"proxy": "landing"}, "stageVariables": null, "path": "/landing", "isBase64Encoded": false}`

Deserialised (and then reserialised) AwsProxyRequest:
`{"body":null,"resource":"/{proxy+}","requestContext":{"resourceId":"123456","apiId":"1234567890","resourcePath":"/{proxy+}","httpMethod":"GET","requestId":"c6af9ac6-7b61-11e6-9a41-93e8deadbeef","extendedRequestId":null,"accountId":"123456789012","identity":{"apiKey":null,"apiKeyId":null,"userArn":null,"cognitoAuthenticationType":null,"caller":null,"userAgent":"Custom User Agent String","user":null,"cognitoIdentityPoolId":null,"cognitoIdentityId":null,"cognitoAuthenticationProvider":null,"sourceIp":"127.0.0.1","accountId":null,"accessKey":null},"authorizer":null,"stage":"prod","path":"/{proxy+}","protocol":null,"requestTime":null,"requestTimeEpoch":0,"elb":null},"multiValueQueryStringParameters":{},"multiValueHeaders":{},"pathParameters":{"proxy":"landing"},"httpMethod":"GET","stageVariables":null,"path":"/landing","requestSource":"API_GATEWAY","base64Encoded":false}
`

It can be seen that my query string values are lost ('someQueryKey') and the headers from the API Gateway also (i.e. 'User-Agent').

It appears this is a regression introduced by: https://github.com/awslabs/aws-serverless-java-container/issues/198#diff-ffda15dbedba00524f29c9961100f4fb
That commit added support for multi value headers and query params, but removed support for single value.

Unfortunately the 'sam local' environment appears to not support multivalues.  See this feature request: https://github.com/awslabs/aws-sam-cli/issues/735
I note that the PR associated with that request ADDs support for multivalues but does not remove the existing functionality for single.

I'm not sure what the best solution is.  Is support for single value being removed from API Gateway?  If so, I can work around this for now by fudging the incoming request and making my own multivalues until support is added for the local test implementation.

If both single and multi need to be supported then https://github.com/awslabs/aws-serverless-java-container/issues/198#diff-ffda15dbedba00524f29c9961100f4fb effectively needs to be reverted and added as a new set of methods instead of a replacement.
I would be happy to take a look at contributing that if it was an appropriate solution.
## Steps to reproduce
See above
## Full log output
See above

**Comment by sapessi** [comment]
Hi Patrick, this is because as of release `1.3` this library only uses API Gateway's new `multiValueHeaders` and `multiValueQueryString` properties. Unfortunately, those are not supported by the SAM CLI yet. I'll push for [https://github.com/awslabs/aws-sam-cli/issues/735](https://github.com/awslabs/aws-sam-cli/issues/735) to be resolved, it is the right thing to do. Apologies for the inconvenience this is causing.

**Comment by patrickherrera** [comment]
No problems. I managed to work around it by parsing the InputStream into Map and also into an AwsProxyRequest object. Then I extracted the headers etc from the Map and added them to the corresponding multivalued version in the AwsProxyRequest and then serialised it back to an InputStream and passed it on.  
It worked, but now I have issues that the local API Gateway is returning back the response from the lambda directly as the raw JSON, rather than extracting the values and setting the response Headers etc and I'm completely stuck.
I'll revisit this down the track.  I was hoping to use a Lambda behind an API Gateway proxy to serve not only some REST endpoints but a simple one page web form.  Dealing with anything besides JSON using the local environment doesn't seem to work very well for me right now.
Thanks
